code/SConstruct

```diff
@@ -1,139 +1,205 @@
-                LINKCOMSTR = "\033[31mLinking\033[0m $TARGET",
-                INSTALLSTR = "\033[33mInstalling\033[0m '$SOURCE' as '$TARGET'",
-
-                # Configuration options
-                TWITTER=False
-                )
+# TODO:
+# - gch precompiled headers
+# - rebuild with PGO after running tests
 
-Export('env')
+Default('sneezy')
 
-# getting some problems with this
-#CacheDir('/mud/build/cache-new')
+Help("""
+Targets: (default is 'sneezy')
+    sneezy          Build the main binary - output: sneezy
+    lowtools        Build the low tools - output: lowtools/*
+    objs/*/*.os     Build a specific file - eg: objs/game/game_hilo.os
+    tags            Rebuild emacs etags
+    -c              Clean up objects and executables (all build output)
+
+Build options and defaults:
+    CXX=        g++     Specify C++ compiler to use
+    CPPPATH=            Add to the header include search path
+    LIBPATH=            Add to the library linker search path
+    olevel=             Level of optimization for the -O flag
+
+Build boolean options: (accept 1/0 y/n yes/no on/off)
+    ccache=     off     Use ccache to cache build object files
+    debug=      off     Enable high-overhead runtime bug detection
+    gprof=      off     Enable gprof profiling
+    harden=     on      Enable low-overhead runtime hardening
+    lto=        off     Optimize entire sneezy binary at once (also sets
+                        optimize=on) -- WARNING: totally wrecks debugging
+    optimize=   off     Use -O3 optimization vs -O1
+    pretty=     on      Prettify output
+    shared=     on      Build MUD code as shared libraries
+    twitter=    off     Enable support for posting shouts to Twitter
+
+Options can be specified on the commandline or via environment variables.
+
+The defaults are optimized for developing with Sneezy rather than for
+deploying it. Good production flags: CXX=g++-6 harden=1 lto=1 shared=0
+(and optimize=1 if you don't need good crash debugging).
+""")
 
-# build the sneezy exe by default, see the help below for other targets
-Default('sneezy')
 
-# go into code and start building
-env.SConscript('code/SConscript',
-               variant_dir='objs',
-               duplicate=0,
-               exports='env')
+flag_ccache =   get_bool_option('ccache',      False)
+flag_debug =    get_bool_option('debug',       False)
+flag_gprof =    get_bool_option('gprof',       False)
+flag_harden =   get_bool_option('harden',      True)
+flag_lto =      get_bool_option('lto',         False)
+flag_optimize = get_bool_option('optimize',    False)
+flag_pretty =   get_bool_option('pretty',      True)
+flag_shared =   get_bool_option('shared',      True)
+flag_twitter =  get_bool_option('twitter',     False)
 
+opt_includes =       get_option('CPPPATH',     '')
+opt_libdirs =        get_option('LIBPATH',     '')
+opt_olevel =         get_option('olevel',      '3' if flag_optimize else '1')
```

Comment:
Hmm, maybe a little, thanks for this poke, apparently there's -Og which is "optimize without breaking debugging".